source/cppfront.cpp

```diff
@@ -3881,50 +4025,8 @@ class cppfront
 
         auto param_type = print_to_string(type_id);
 
-        //  If there are template parameters on this function or its enclosing
-        //  type, see if this parameter's name is an unqualified-id with a
-        //  template parameter name, or mentions a template parameter as a
-        //  template argument
-        auto has_template_parameter_type_named = [](
-            declaration_node const& decl,
-            std::string_view        name
-            )
-            -> bool
-        {
-            if (decl.template_parameters) {
-                for (auto& tparam : decl.template_parameters->parameters)
-                {
-                    assert(
-                        tparam
-                        && tparam->name()
-                    );
-                    //  For now just do a quick string match
-                    auto tparam_name = tparam->name()->to_string(true);
-                    if (
-                        tparam->declaration->is_type()
-                        && (
-                            name == tparam_name
-                            || name.find("<"+tparam_name) != std::string_view::npos
-                            || name.find(","+tparam_name) != std::string_view::npos
-                        )
-                        )
-                    {
-                        return true;
-                    }
-                }
-            }
-            return false;
-        };
-
         assert( current_declarations.back() );
-        auto is_dependent_parameter_type =
-            has_template_parameter_type_named( *current_declarations.back(), param_type )
-            || (
-                current_declarations.back()->parent_is_type()
-                && current_declarations.back()->has_name("operator=")
-                && has_template_parameter_type_named( *current_declarations.back()->get_parent(), param_type)
-                )
-            ;
+        auto is_dependent_parameter_type = current_declarations.back()->is_dependent_parameter_type( param_type );
```

Comment:
The reason for #534 is because this line asks an incomplete question.
In addition to asking "is this parameter's type dependent",
it should also be asking "is this parameter's type deducible".

When emitting `typename`, I'm also asking "is this type dependent", which is also incomplete.
That's why `std::type_identity_t<T>` ends up prefixed with `typename`.
In addition, I should also be asking "does this need to be prefixed with `typename`",
which requires that a non-terminal component of the _qualified-id_ is a dependent _template-id_.